interfaces/repo.go

```diff
@@ -267,29 +267,66 @@ func (r *Repository) RemoveSlot(snapName, slotName string) error {
 
 // Connect establishes a connection between a plug and a slot.
 // The plug and the slot must have the same interface.
-func (r *Repository) Connect(plugSnapName, plugName, slotSnapName, slotName string) error {
+func (r *Repository) Connect(plugSnapName, plugName, slotSnapName, slotName string) (*PlugRef, *SlotRef, error) {
 	r.m.Lock()
 	defer r.m.Unlock()
 
 	// Ensure that such plug exists
 	plug := r.plugs[plugSnapName][plugName]
 	if plug == nil {
-		return fmt.Errorf("cannot connect plug %q from snap %q, no such plug", plugName, plugSnapName)
+		return nil, nil, fmt.Errorf("cannot connect plug %q from snap %q, no such plug", plugName, plugSnapName)
+	}
+
+	var slot *Slot
+	// connect the plug to slot with matching interface if slot is omitted
+	if slotName == "" {
+		// connect OS snap if snap name is omitted
+		if slotSnapName == "" {
+			for _, slots := range r.slots {
```

Comment:
Instead of having two loops that look alike, I suggest doing a loop here with the only goal of finding the slotSnapName of the core snap, and then falling through to the logic below (currently under the else), which is common between the two.